subsys/bluetooth/mesh/access.c

```diff
@@ -2614,3 +2634,220 @@ uint8_t bt_mesh_comp_parse_page(struct net_buf_simple *buf)
 
 	return page;
 }
+
+#if defined CONFIG_BT_MESH_ACCESS_TX_MSG_RANDOMIZER
+
+struct access_message_chunk {
+	sys_snode_t node;
+	uint8_t data[CONFIG_BT_MESH_DELAYED_ACCESS_MSG_CHUNK_SIZE];
+};
+
+static void randomizer_handler(struct k_work *w);
+
+static struct access_message {
+	bool busy;
+	struct bt_mesh_msg_ctx ctx;
+	const struct bt_mesh_model *model;
+	const struct bt_mesh_send_cb *cb;
+	void *cb_data;
+	uint32_t fired_time;
+	size_t len;
+	sys_slist_t chunks;
+} access_messages[CONFIG_BT_MESH_DELAYED_ACCESS_MSG_COUNT];
+
+static struct {
+	struct k_work_delayable random_delay;
+	struct access_message *pending_msg;
+} access_randomizer = {.random_delay = Z_WORK_DELAYABLE_INITIALIZER(randomizer_handler)};
+
+K_MEM_SLAB_DEFINE_STATIC(access_message_chunks, sizeof(struct access_message_chunk),
+			 CONFIG_BT_MESH_DELAYED_ACCESS_MSG_CHUNK_COUNT, 4);
+
+static void reschedule_randomizer(void)
+{
+	uint32_t fired_time_min = 0;
+	uint32_t curr_time;
+	k_timeout_t delay = K_NO_WAIT;
+
+	access_randomizer.pending_msg = NULL;
+
+	for (int i = 0; i < CONFIG_BT_MESH_DELAYED_ACCESS_MSG_COUNT; i++) {
+		if (access_messages[i].busy) {
+			if (fired_time_min < access_messages[i].fired_time) {
+				fired_time_min = access_messages[i].fired_time;
+				access_randomizer.pending_msg = &access_messages[i];
+			}
+		}
+	}
```

Comment:
Why not to use linked list here for sorting messages in order they need to be sent? Then the head always points to the next message that needs to be sent.